File: src/db/queries.py
import logging
from sqlalchemy import text
from sqlalchemy.exc import DBAPIError
from sqlalchemy.ext.asyncio import AsyncSession


logger = logging.getLogger(__name__)


async def get_all_fhir_by_patient(db: AsyncSession, patient_id: str) -> list[dict]:
    """Fetch all FHIR resources for a patient. Returns id, resource_type, resource, received_at."""

    logger.info("fhir_query | patient_id=%s | filter=all", patient_id)
    result = await db.execute(
        text(
            """
            SELECT id AS resource_id, resource_type, resource, received_at
            FROM fhir_resources
            WHERE patient_id = :pid
            ORDER BY received_at
            """
        ),
        {"pid": patient_id},
    )
    rows = [
        {
            "resource_id": str(r.resource_id),
            "resource_type": r.resource_type,
            "resource": r.resource,
            "received_at": str(r.received_at) if r.received_at else "",
        }
        for r in result.fetchall()
    ]
    logger.info(
        "fhir_query_complete | patient_id=%s | row_count=%d", patient_id, len(rows)
    )
    return rows


async def get_patient_overview(db: AsyncSession, patient_id: str) -> dict:
    """Lightweight overview: resource type counts and date ranges. No clinical content."""

    logger.info("get_patient_overview | patient_id=%s", patient_id)
    try:
        result = await db.execute(
            text(
                """
                SELECT
                    resource_type,
                    COUNT(*) AS count,
                    MIN(received_at) AS min_date,
                    MAX(received_at) AS max_date
                FROM fhir_resources
                WHERE patient_id = :pid
                GROUP BY resource_type
                ORDER BY count DESC
                """
            ),
            {"pid": patient_id},
        )
        rows = result.fetchall()

    except DBAPIError as e:
        logger.error("get_patient_overview_failed | patient_id=%s | error=%s", patient_id, str(e))
        raise ValueError("Invalid patient_id: must be a valid UUID.") from e

    except Exception as e:
        logger.error("get_patient_overview_failed | patient_id=%s | error=%s", patient_id, str(e))
        raise e
    
    if not rows:
        raise ValueError(f"No resources found for patient {patient_id}. Patient does not exist!")
    
    overview = {
        "by_type": [
            {
                "resource_type": r.resource_type,
                "count": r.count,
                "min_date": str(r.min_date) if r.min_date else None,
                "max_date": str(r.max_date) if r.max_date else None,
            }
            for r in rows
        ],
        "total_resources": sum(r.count for r in rows),
    }
    logger.info("get_patient_overview_complete | patient_id=%s | types=%d", patient_id, len(rows))
    return overview


async def get_fhir_by_type(
    db: AsyncSession,
    patient_id: str,
    resource_type: str,
    limit: int = 20,
) -> list[dict]:
    """Fetch FHIR resources for a patient filtered by type (e.g., 'Condition', 'Observation')."""

    logger.info("fhir_query | patient_id=%s | filter=%s | limit=%d", patient_id, resource_type, limit)
    try:
        async with db.begin_nested():
            result = await db.execute(
                text(
                    """
                    SELECT id AS resource_id, resource_type, resource, received_at
                    FROM fhir_resources
                    WHERE patient_id = :pid AND resource_type = :rt
                    ORDER BY received_at
                    LIMIT :lim
                    """
                ),
                {"pid": patient_id, "rt": resource_type, "lim": limit},
            )
            rows = result.fetchall()
    except Exception as e:
        logger.warning(
            "fhir_query_failed | patient_id=%s | filter=%s | error=%s",
            patient_id, resource_type, str(e),
        )
        raise ValueError(f"Failed to get FHIR resources by type {resource_type} for patient {patient_id}: {str(e)}")

    out = [
        {
            "resource_id": str(r.resource_id),
            "resource_type": r.resource_type,
            "resource": r.resource,
            "received_at": str(r.received_at) if r.received_at else "",
        }
        for r in rows
    ]
    logger.info(
        "fhir_query_complete | patient_id=%s | row_count=%d", patient_id, len(out)
    )
    return out


async def search_resources_by_keyword(
    db: AsyncSession,
    patient_id: str,
    keyword: str,
    limit: int = 10,
) -> list[dict]:
    """Search FHIR resources by keyword in JSON content (ILIKE)."""

    logger.info("fhir_search | patient_id=%s | keyword=%s | limit=%d", patient_id, keyword, limit)
    pattern = f"%{keyword}%"
    try:
        async with db.begin_nested():
            result = await db.execute(
                text(
                    """
                    SELECT id AS resource_id, resource_type, resource, received_at
                    FROM fhir_resources
                    WHERE patient_id = :pid AND resource::text ILIKE :pat
                    ORDER BY received_at
                    LIMIT :lim
                    """
                ),
                {"pid": patient_id, "pat": pattern, "lim": limit},
            )
            rows = result.fetchall()
    except Exception as e:
        logger.warning(
            "fhir_search_failed | patient_id=%s | keyword=%s | error=%s",
            patient_id, keyword, str(e),
        )
        raise ValueError(f"Failed to search resources by keyword {keyword} for patient {patient_id}: {str(e)}")

    out = [
        {
            "resource_id": str(r.resource_id),
            "resource_type": r.resource_type,
            "resource": r.resource,
            "received_at": str(r.received_at) if r.received_at else "",
        }
        for r in rows
    ]
    logger.info(
        "fhir_search_complete | patient_id=%s | row_count=%d", patient_id, len(out)
    )
    return out


async def execute_raw_sql(
    db: AsyncSession,
    sql: str,
    params: dict,
) -> list[dict]:
    """Execute validated SQL with bound parameters."""
    logger.info("sql_execute | sql_preview=%s", sql)
    try:
        async with db.begin_nested():
            result = await db.execute(text(sql), params)
            rows = result.fetchall()
    except Exception as e:
        logger.warning("sql_execute_failed | error=%s | sql=%s", str(e), sql)
        raise ValueError(f"Failed to execute raw SQL: {str(e)}")

    out = []
    for r in rows:
        d = dict(r._mapping)
        for k, v in d.items():
            if v is not None and hasattr(v, "isoformat"):
                d[k] = v.isoformat()
            elif v is not None and not isinstance(v, (str, int, float, bool)):
                d[k] = str(v)
        out.append(d)
    return out


async def get_fhir_resources_schema_info(db: AsyncSession) -> dict:
    """Return column names and types for fhir_resources and fhir_note_text."""

    try:
        result = await db.execute(
            text(
                """
                SELECT table_name, column_name, data_type
                FROM information_schema.columns
                WHERE table_schema = 'public'
                AND table_name IN ('fhir_resources')
                ORDER BY table_name, ordinal_position
                """
            ),
        )
        schema: dict[str, list[dict]] = {}

        if not result.fetchall():
            raise ValueError("No schema found for 'fhir_resources'.")
        
        for r in result.fetchall():
            table = r.table_name
            if table not in schema:
                schema[table] = []
                schema[table].append({"column": r.column_name, "type": r.data_type})
            return schema
        
    except Exception as e:
        logger.error("get_fhir_resources_schema_info_failed | error=%s", str(e))
        raise e

File: ts-version/src/db/queries.ts
import { logger } from '../config/logging';

function formatError(error: unknown) {
  if (error instanceof Error) {
    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
      ...(error as any),
    };
  }
  return { message: String(error) };
}

export interface DatabasePool {
  query(text: string, params?: any[]): Promise<any>;
  end(): Promise<void>;
}

export async function getAllFHIRByPatient(db: DatabasePool, patientId: string): Promise<any[]> {
  logger.info(`fhir_query | patient_id=${patientId} | filter=all`);
  
  const result = await db.query(`
    SELECT id AS resource_id, resource_type, resource, received_at
    FROM fhir_resources
    WHERE patient_id = $1
    ORDER BY received_at
  `, [patientId]);

  const rows = result.rows.map((r: any) => ({
    resource_id: String(r.resource_id),
    resource_type: r.resource_type,
    resource: r.resource,
    received_at: r.received_at ? String(r.received_at) : "",
  }));

  logger.info(`getAllFHIRByPatient completed | patient_id=${patientId} | row_count=${rows.length}`);
  return rows;
}

export async function getPatientOverview(db: DatabasePool, patientId: string): Promise<any> {
  logger.info(`fhir_overview | patient_id=${patientId}`);
  
  const result = await db.query(`
    SELECT
      resource_type,
      COUNT(*) AS count,
      MIN(received_at) AS min_date,
      MAX(received_at) AS max_date
    FROM fhir_resources
    WHERE patient_id = $1
    GROUP BY resource_type
    ORDER BY count DESC
  `, [patientId]);

  const rows = result.rows;
  const overview = {
    by_type: rows.map((r: any) => ({
      resource_type: r.resource_type,
      count: r.count,
      min_date: r.min_date ? String(r.min_date) : null,
      max_date: r.max_date ? String(r.max_date) : null,
    })),
    total_resources: rows.reduce((sum: number, r: any) => sum + r.count, 0),
  };

  logger.info(`fhir_overview_complete | patient_id=${patientId} | types=${rows.length}`);
  return overview;
}

export async function getFHIRByType(
  db: DatabasePool,
  patientId: string,
  resourceType: string,
  limit: number = 20
): Promise<any[]> {
  logger.info(`fhir_query | patient_id=${patientId} | filter=${resourceType} | limit=${limit}`);
  
  try {
    const result = await db.query(`
      SELECT id AS resource_id, resource_type, resource, received_at
      FROM fhir_resources
      WHERE patient_id = $1 AND resource_type = $2
      ORDER BY received_at
      LIMIT $3
    `, [patientId, resourceType, limit]);

    const rows = result.rows.map((r: any) => ({
      resource_id: String(r.resource_id),
      resource_type: r.resource_type,
      resource: r.resource,
      received_at: r.received_at ? String(r.received_at) : "",
    }));

    logger.info(`getFHIRByType completed | patient_id=${patientId} | row_count=${rows.length}`);
    return rows;
  } catch (error) {
    logger.warn(
      `fhir_query_failed | patient_id=${patientId} | filter=${resourceType} | error=${String(error)}`
    );
    throw error;
  }
}

export async function searchResourcesByKeyword(
  db: DatabasePool,
  patientId: string,
  keyword: string,
  limit: number = 10
): Promise<any[]> {
  logger.info(`fhir_search | patient_id=${patientId} | keyword=${keyword} | limit=${limit}`);
  const pattern = `%${keyword}%`;
  
  try {
    const result = await db.query(`
      SELECT id AS resource_id, resource_type, resource, received_at
      FROM fhir_resources
      WHERE patient_id = $1 AND resource::text ILIKE $2
      ORDER BY received_at
      LIMIT $3
    `, [patientId, pattern, limit]);

    const rows = result.rows.map((r: any) => ({
      resource_id: String(r.resource_id),
      resource_type: r.resource_type,
      resource: r.resource,
      received_at: r.received_at ? String(r.received_at) : "",
    }));

    logger.info(`fhir_search_complete | patient_id=${patientId} | row_count=${rows.length}`);
    return rows;
  } catch (error) {
    logger.warn(
      `fhir_search_failed | patient_id=${patientId} | keyword=${keyword} | error=${String(error)}`
    );
    throw error;
  }
}

export async function executeRawSQL(
  db: DatabasePool,
  sql: string,
  params: any
): Promise<any[]> {
  logger.info(`sql_execute | sql_preview=${sql}`);
  
  try {
    const result = await db.query(sql, params);
    const rows = result.rows;

    const out = rows.map((r: any) => {
      const d = { ...r };
      for (const [k, v] of Object.entries(d)) {
        if (v !== null && typeof v === 'object' && 'toISOString' in v) {
          d[k] = (v as Date).toISOString();
        } else if (v !== null && !['string', 'number', 'boolean'].includes(typeof v)) {
          d[k] = String(v);
        }
      }
      return d;
    });

    return out;
  } catch (error) {
    logger.warn('sql_execute_failed', { sql, error: formatError(error) });
    throw error;
  }
}

export async function getFHIRResourcesSchemaInfo(db: DatabasePool): Promise<any> {
  const result = await db.query(`
    SELECT table_name, column_name, data_type
    FROM information_schema.columns
    WHERE table_schema = 'public'
    AND table_name IN ('fhir_resources')
    ORDER BY table_name, ordinal_position
  `);

  const schema: { [key: string]: any[] } = {};
  for (const r of result.rows) {
    const table = r.table_name;
    if (!schema[table]) {
      schema[table] = [];
    }
    schema[table].push({ column: r.column_name, type: r.data_type });
  }
  
  return schema;
}


// Get patient record by profile ID with clear_verifications
export const getPatientByProfileId = async (db: DatabasePool, profileId: number) => {
  const query = `
    SELECT 
      p.patient_id,
      p.profile_id,
      p.current_verification_id,
      p.created_at as patient_created_at,
      p.updated_at as patient_updated_at,
      cv.verification_id,
      cv.clear_identity_token,
      cv.claims,
      cv.verified_at,
      cv.verification_status,
      cv.verification_level,
      cv.created_at as verification_created_at
    FROM patients p
    LEFT JOIN clear_verifications cv ON p.current_verification_id = cv.verification_id
    WHERE p.profile_id = $1
  `;
  const result = await db.query(query, [profileId]);
  return result.rows[0];
};


// Get FHIR data for specific patient
export const getFHIRUISummaryDataByPatientId = async (db: DatabasePool, patientId: string) => {
  // TODO: Replace with your provided query that produces same structure as fhir.json
  const query = "SELECT get_patient_summary($1, true) as fhir_data"
  const result = await db.query(query, [patientId]);
  return result.rows[0]?.fhir_data || {};
};